File: dotnet/Ciel.Api/Domain/PostVisibility.cs
using System.Text.Json.Serialization;

namespace Ciel.Api.Domain;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum PostVisibility
{
    Public,
    FollowersOnly,
}

public static class PostVisibilityDb
{
    public static PostVisibility? FromDb(string value) => value switch
    {
        "public" => PostVisibility.Public,
        "followers_only" => PostVisibility.FollowersOnly,
        _ => null,
    };

    public static string AsDb(PostVisibility visibility) => visibility switch
    {
        PostVisibility.Public => "public",
        PostVisibility.FollowersOnly => "followers_only",
        _ => throw new ArgumentOutOfRangeException(nameof(visibility)),
    };
}

File: spring/src/main/java/com/ciel/domain/PostVisibility.java
package com.ciel.domain;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

/**
 * Mirrors Rust's {@code domain::post::PostVisibility}. JSON representation
 * follows the Rust serde enum names verbatim ({@code Public}, {@code FollowersOnly});
 * the DB representation is lowercase snake_case ({@code public}, {@code followers_only}).
 */
public enum PostVisibility {
    PUBLIC("public", "Public"),
    FOLLOWERS_ONLY("followers_only", "FollowersOnly");

    private final String db;
    private final String json;

    PostVisibility(String db, String json) {
        this.db = db;
        this.json = json;
    }

    @JsonValue
    public String jsonName() {
        return json;
    }

    public String asDb() {
        return db;
    }

    public static PostVisibility fromDb(String value) {
        if (value == null) {
            return null;
        }
        return switch (value) {
            case "public" -> PUBLIC;
            case "followers_only" -> FOLLOWERS_ONLY;
            default -> null;
        };
    }

    @JsonCreator
    public static PostVisibility fromJson(String value) {
        for (PostVisibility v : values()) {
            if (v.json.equals(value)) {
                return v;
            }
        }
        throw new IllegalArgumentException("unknown post visibility: " + value);
    }
}
